### PR TITLE
Writing flow: restore click redirect between blocks

### LINKED ISSUE #28682: Block hit area regression
## Description
The area around blocks that can be clicked to trigger selection seems to have shifted upward. That is, if you click just below the end of the visible block content, the next block is selected, whereas before that would select the block above (closest to the click). 

In the screen recordings, I've demoed the issue with a Separator block but it seems to affect all blocks. The first video is to show the expected selection behavior from Guteberg included in WP5.6. It can be seen a click well below the line still selects the Separator block. The second video shows that a click on the line itself appears not to work* and that the click has to be above the line in order to select the Separator.

*The appearance that a click on the line does not work is likely due to the location of the hotspot in the macOS cursor.

## Step-by-step reproduction instructions
1. In a block editor running the latest Gutenberg from master try clicking just below the visible content of a block and observe which adjacent block is selected. 
2. Try the same on the version of Gutenberg built into WP 5.6 and observe the difference.

## Expected behaviour
On a click between blocks, the block closest to the click is selected

## Actual behaviour
On a click between blocks, the block following the block that was closest to the click is selected.

### Gutenberg on WP 5.6

https://user-images.githubusercontent.com/9000376/106695950-84a47b00-6590-11eb-9098-8d3c0c43d7d7.mp4

### Gutenberg plugin built from e4eb4be3a9

https://user-images.githubusercontent.com/9000376/106695956-8a9a5c00-6590-11eb-86ea-6e91b8c2ac61.mp4

## WordPress information
- WordPress version: 5.6.0
- Gutenberg version: 9.8.0 (I think, it's present since at least e4eb4be3a9)
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

## Device information
- Device: Desktop
- Operating system: macOS 10.12.6
- Browser: Chrome 88.0

### LINKED ISSUE #32129: Separator block is difficult to select due to small size
## What problem does this address?

This feedback came in as part of the [sixth call for testing for the FSE Outreach Program](https://make.wordpress.org/test/2021/05/12/fse-program-testing-call-6-stick-the-landing-pages/#comment-1612): 

> I couldn’t select the Separator block, however hard I tried. I had to resort to using the List View.

Here's a quick video showing a bit how precise one needs to be in order to select it: 

https://user-images.githubusercontent.com/26996883/119248015-68324b80-bb4b-11eb-9469-ee925baca9f3.mov

## What is your proposed solution?

It's unclear to me how this has been handled previously but the only option that comes to mind is expanding the "surface area" for selecting this block.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #32129.
Fixes #28682.
Restored a feature removed in #27860 where a click between two block would set focus/caret to the closest point.

I rewrote the feature to be its own hook, independent of the insertion point.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/components/inner-blocks/index.js [hunk 1]
```diff
@@ -157,6 +158,7 @@ export function useInnerBlocksProps( props = {}, options = {} ) {
 
 	const ref = useMergeRefs( [
 		props.ref,
+		useInBetweenClickRedirect(),
```
Comment (ellatrix) [inline packages/block-editor/src/components/inner-blocks/index.js]: Ideally the block list root should just be reusing the inner blocks hooks. See #29895.

File: packages/e2e-tests/specs/editor/various/__snapshots__/writing-flow.test.js.snap [hunk 2]
```diff
@@ -278,6 +288,12 @@ exports[`Writing Flow should not prematurely multi-select 1`] = `
 <!-- /wp:paragraph -->"
 `;
 
+exports[`Writing Flow should only consider the content as one tab stop 1`] = `
+"<!-- wp:table -->
+<figure class=\\"wp-block-table\\"><table><tbody><tr><td></td><td>2</td></tr><tr><td></td><td></td></tr></tbody></table></figure>
+<!-- /wp:table -->"
+`;
```
Comment (ellatrix) [inline packages/e2e-tests/specs/editor/various/__snapshots__/writing-flow.test.js.snap:301]: This snapshot has just been moved to alphabetical order automatically.

**Comment by stokesman** [comment]
I tested with gutenberg.run and it does indeed resolve the linked issues.

**Comment by stokesman** [comment]
Great work on this! The latest commits have enhanced the selection experience. I still found a couple little oddities but they don't look like blockers to me as they are in trunk as well. Highlighting them for the record and in case a solution occurs. 

https://user-images.githubusercontent.com/9000376/121917250-cc2ed680-cce9-11eb-9c16-18bc117b10e7.mp4

Okay, that first one isn't exactly in trunk but I think it would be if a horizontally between click actually attempted to select a child.

It so happens I'd retested this a few days ago when I got curious if it'd done anything for horizontal blocks and had noticed some issues, but your commits fixed them before I got around to highlighting them.

**Comment by ellatrix** [comment]
@stokesman There was a bug with inner blocks. Could you retest? :)

**Comment by stokesman** [comment]
I retested but couldn't spot any differences. I don't know if ed43e6f was meant to address it but regarding that "outside clicks select last-child" I have a few more details that might help (but it is present in trunk so I don't think it needs fixed here).
- I've only found it happening with Navigation and Buttons, not with Columns or Social Icons (I'm not sure if there are other horizontal blocks to test)
- If the last-child is already selected an outside click unselects it 
  - subsequent outside clicks do nothing
  - except double-clicks which select the last child
 
https://user-images.githubusercontent.com/9000376/122262708-88bf9e00-ce8a-11eb-9e70-0b86d32adaab.mp4

**Comment by ellatrix** [comment]
@stokesman I'm not sure if I can reproduce it consistently and I'm not sure if it's related to this PR. Does it happen in trunk too? Anyway, I adjusted the PR to only look in the orientation of the block list, which should help reduce issues. Tests are passing now and it seems to work well.

**Comment by stokesman** [comment]
Yes, anything I last flagged was also present in trunk*. I've retested this and it works very well. LGTM as in Lets Get This Merged!

*Having retested, I found the odd behavior I flagged earlier about clicks outside (Buttons or Navigation) block is still present in both trunk and this branch but it has changed the same way in both (now it selects the first-child) and that seems to further indicate it is unrelated to any changes in this branch.

**Comment by ellatrix** [comment]
@stokesman Maybe the behaviour you're experiencing in related to #34009?

**Comment by stokesman** [comment]
> Maybe the behaviour you're experiencing in related to #34009?

Seems to be it.

**Comment by paaljoachim** [comment]
Hi

What are the next steps in this PR?
Thanks!

**Comment by ellatrix** [comment]
Simplified this in #42475.

**Pull request closed without merging**